Extended Tour Documentation (#614)
open-component-model/ocm: examples/lib/tour/docsrc/05-transporting-component-versions/README.md

# Transporting Component Versions

This [tour](example.go) illustrates the basic support for
transporting content from one environment into another.

## Running the example

You can just call the main program with some config file option (`--config <file>`).
The config file should have the following content:

```yaml
repository: ghcr.io/mandelsoft/ocm
targetRepository:
  type: CommonTransportFormat
  filePath: /tmp/example05.target.ctf
  fileFormat: directory
  accessMode: 2
username:
password:
```

Any supported kind of target repository can be specified by using its
specification type. An OCI regisztry target would look like this:

```yaml
repository: ghcr.io/mandelsoft/ocm
username:
password:
targetRepository:
  type: OCIRegistry
  baseUrl: ghcr.io/mandelsoft/targetocm
ocmConfig: <config file>
```

> The actual version of the example just works with the filesystem 

Reviewer: ```suggestion
The actual version of the example just works with the file system 
```

Author: done

target, because it is not possible to specify credentials for the

> target repository in this simple config file. But, if you specific an [OCM config file](../04-working-with-config/README.md) you can

Reviewer: ```suggestion
target repository in this simple config file. But, if you specify
an [OCM config file](../04-working-with-config/README.md) you can
```

Author: done

> add more credential settings to make target repositories possible

Reviewer: was meinst du damit? "to make target repositories possible requiring credentials" ?

Author: explained

```

For our source we just use the component version provided by the last examples
in a remote repository.
Therefore, we set up the credentials context, again, as has
been shown in [tour 03]({{using-cred-management}}).

```go
{{include}{../../05-transporting-component-versions/example.go}{set credentials}}
```

For the transport, we first get access to the component version
we want to transport, by getting the source repository and looking up
the desired component version.

```go
{{include}{../../05-transporting-component-versions/example.go}{source}}
```

We could just add this version to the target repository, but this
would not be a real transport, just a copy of the component descriptor
and the local resources. Transport potentially means more, all the
described artifacts should be copied into the target environment, also.

Such an action is done by a library function `transfer.Transfer`.
It takes several settings influencing the transport mode
(for example transitive or value transport).
Here, all resources are transported per value, all external
references will be inlined as `localBlob`s and imported into
the target environment, applying blob upload handlers

> where possible. For a CTF Archive as target, there are no

Reviewer: ```suggestion
where possible. For a CTF archive as target, there are no
```

Author: done

> configured handlers, by default. Therefore, all resources will

Reviewer: ```suggestion
configured handlers by default. Therefore, all resources will
```

Author: done
